wp_api/src/authenticator.rs

```diff
@@ -0,0 +1,304 @@
+use http::header::HeaderMap;
+use http::header::HeaderValue;
+use std::sync::{Arc, RwLock};
+
+use crate::{
+    request::{
+        endpoint::ApiBaseUrl, RequestExecutor, RequestMethod, WpNetworkHeaderMap, WpNetworkRequest,
+        WpNetworkRequestBody, WpNetworkResponse,
+    },
+    RequestExecutionError, WpLoginCredentials,
+};
+
+const CONTENT_TYPE_FORM: &str = "application/x-www-form-urlencoded";
+
+#[derive(Debug)]
+pub enum AuthenticationError {
+    ReAuthenticationNotApplicable,
+    IncorrectCredentials,
+    RequestUrlDoesNotMatchSite,
+    InvalidWebFormContent,
+}
+
+#[async_trait::async_trait]
+pub trait Authenticator: Send + Sync + std::fmt::Debug {
+    async fn authenticate(
+        &self,
+        request: &WpNetworkRequest,
+    ) -> Result<HeaderMap, AuthenticationError>;
+
+    fn should_reauthenticate(&self, response: &WpNetworkResponse) -> bool {
```

Comment:
```suggestion
    fn should_reauthenticate(&self, response_status_code: u16) -> bool {
```

I think we should limit the function signature as much as possible. For now, sharing the status code should be enough. In the future, we might need to get a little more sophisticated, but doing this should help discourage ourselves from doing any unnecessary processing of the response. For example, we wouldn't want this code to parse the response as that'll potentially duplicating the parsing logic.